datafusion/functions/src/core/arrow_cast.rs

```diff
@@ -87,6 +89,10 @@ impl ScalarUDFImpl for ArrowCastFunc {
         internal_err!("arrow_cast should return type from exprs")
     }
 
+    fn is_nullable(&self, args: &[Expr], schema: &dyn ExprSchema) -> bool {
+        args.iter().any(|e| e.nullable(schema).ok().unwrap_or(true))
```

Comment:
This was added due to the test from #12050

Reply:
It seems there are more functions required to customize nullable based on input's nullability. But we can add them gradually